sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -205,10 +205,14 @@ object ConstantPropagation extends Rule[LogicalPlan] {
     def replaceConstants0(expression: Expression) = expression transform {
       case a: AttributeReference => constantsMap.getOrElse(a, a)
     }
-    condition transform {
-      case e @ EqualTo(_, _) if !predicates.contains(e) => replaceConstants0(e)
-      case e @ EqualNullSafe(_, _) if !predicates.contains(e) => replaceConstants0(e)
-    }
+    splitConjunctivePredicates(condition).map {
+      // Don't replace constants in IsNotNull because InferFiltersFromConstraints may infer
+      // another IsNotNull
+      case isNotNull: IsNotNull => isNotNull
+      case e: EqualTo if predicates.contains(e) => e
```

Comment:
Now we don't check `if predicates.contains(e)` for nested `EqualTo`. is this a bug?

Reply:
This is a improvement, For example:
```
(a > b + 1 && b = 1 || c = 1) && (b = 1) ===> (a > 2 || c = 1) && (b = 1)
```
It cannot be optimized before.

Reply:
I get your point. I think we should rewrite the code to make the intention more clear: we can't replace attr with constant if the attr is in the source equal to predicates. We can set a tree node tag in the source equal to predicates in `traverse`, and skip these source equal to predicates when optimizing by checking the tree node tag.

Reply:
OK

Reply:
It looks a little complicated. We also need to clean up the tag:
https://github.com/apache/spark/pull/36850/commits/1ee604ceba2369625195a44e41ed6b696edd3a34